cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/NotificationSubscriberService.java

```diff
@@ -0,0 +1,341 @@
+    }
+    if (taskExecutorService != null) {
+      taskExecutorService.shutdownNow();
+    }
+    LOG.info("NotificationSubscriberService stopped.");
+  }
+
+  private class NotificationSubscriberThread extends Thread {
+    private final String topic;
+    private final RetryStrategy scheduleStrategy;
+    private final MultiThreadMessagingContext messagingContext;
+    private int emptyFetchCount;
+    private int failureCount;
+    private String messageId;
+
+    NotificationSubscriberThread(String topic, @Nullable String messageId) {
+      super(String.format("NotificationSubscriberThread-%s", topic));
+      this.topic = topic;
+      this.messageId = messageId;
+      // Retry with delay ranging from 0.1s to 30s
+      scheduleStrategy =
+        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+      this.messagingContext = new MultiThreadMessagingContext(messagingService);
+      emptyFetchCount = 0;
+      failureCount = 0;
+    }
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = fetchNewNotifications();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // It's triggered by stop
+          Thread.currentThread().interrupt();
+        }
+      }
+    }
+
+    /**
+     * Fetch new notifications and update job queue
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long fetchNewNotifications() {
+      try {
+        MessageFetcher fetcher = messagingContext.getMessageFetcher();
+        emptyFetchCount++;
+        try (CloseableIterator<Message> iterator = fetcher.fetch(NamespaceId.DEFAULT.getNamespace(),
+                                                                 topic, 100, messageId)) {
+          LOG.trace("Fetch with messageId = {}", messageId);
+          String currentMessageId = null;
+          try {
+            while (iterator.hasNext() && isRunning()) {
+              emptyFetchCount = 0;
+              Message message = iterator.next();
+              Notification notification;
+              try {
+                notification = GSON.fromJson(new String(message.getPayload(), StandardCharsets.UTF_8),
+                                             Notification.class);
+                processNotification(notification);
+              } catch (JsonSyntaxException e) {
+                LOG.warn("Failed to decode message with id {}. Skipped. ", message.getId(), e);
+              }
+              // Record current message's Id no matter decode is successful or not,
+              // so that this message will be skipped in next fetch
+              currentMessageId = message.getId();
+            }
+          } catch (Exception e) {
+            LOG.warn("Failed to fetch new notification. Will retry in next run", e);
```

Comment:
This will log a lot of messages when master start before TMS is available. You should catch the specific exception and log accordingly. E.g. if the exception is `ServiceUnavailableException` or `TopicNotFoundException`, you don't need to log or don't need to log right away, but only log when it happens multiple times in a a period of time. You might need a new `LogSampler` to do so.